### PR TITLE
initial module addition

### PR DESCRIPTION
**Pull request by milesmatthias**
Initial addition of the module, along with explanation & examples, but a few items left to do:

* [x] figure out how to properly test this, since this module does not create any Terraform or GCP resources

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,80 @@
...
+  description = "Short name of the given region"
+  value       = "${module.utils.region_short_name}"
+}
+```
+
+Functional examples are included in the
+[examples](./examples/) directory.
+
+<!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
+## Inputs
```
Comment (morgante) [inline README.md:37]: These inputs don't look correct?

File: examples/simple_example/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Simple Example
+
+This example illustrates how to use the `utils` module.
+
+<!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
+## Inputs
+
+| Name | Description | Type | Default | Required |
+|------|-------------|:----:|:-----:|:-----:|
+| bucket\_name | The name of the bucket to create. | string | n/a | yes |
```
Comment (morgante) [inline examples/simple_example/README.md]: Please update docs.

File: examples/simple_example/main.tf [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+
+provider "google" {
+  version = "~> 2.0"
+}
+
+module "utils" {
+  source  = "../.."
+  region  = "${var.region}"
```
Comment (morgante) [inline examples/simple_example/main.tf]: Examples don't need to be excessively parameterized. In this case, let's just hardcode a single region.

If we also want to exemplify/share the multi-regional case, do a separate (hardcoded) example for that.

File: test/setup/make_source.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
```
Comment (morgante) [inline test/setup/make_source.sh]: If you update to the latest container, this file is no longer needed: https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/blob/master/infra/build/Makefile#L27

File: variables.tf [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "regions" {
+  description = "A list of GCP region codes for each of the given regions. i.e. [\"us-central1\"] => {\"us-central1\"=\"usc1\"}"
```
Comment (morgante) [inline variables.tf]: Isn't this just a list of GCP region codes to compute.

For what it's worth, I'm not sure this module really needs inputs at all? We could simply output all the regions each time.

**Review by morgante — CHANGES_REQUESTED** [review]

File: variables.tf (hunk 5 above)
Comment (milesmatthias) [reply variables.tf]: Is there a change you'd like me to make here? Or an issue for later I should make?

**Comment by milesmatthias** [comment]
The test pass locally, but not sure why they're not passing on cloudbuild with the new devtools image...

**Comment by morgante** [comment]
@milesmatthias It's failing because you have nothing in `test/setup`. You need *some* Terraform config in there.

Since this module doesn't actually need a project or credentials to run, this could be as simple as a blank Terraform file or a random_pet resource.

File: variables.tf (hunk 5 above)
Comment (morgante) [reply variables.tf]: I'd like you to make the change here (remove the `regions` input.)

File: README.md [hunk 6]
```diff
@@ -1 +1,81 @@
...
+
+## Usage
+
+Basic usage of this module is as follows:
+
+```hcl
+module "utils" {
+  source  = "terraform-google-modules/utils/google"
+  version = "~> 0.1"
+  regions = ["us-central1", "us-west1"]
```
Comment (morgante) [inline README.md]: Let's remove the "regions" input.

**Review by morgante — CHANGES_REQUESTED** [review]
Just need to remove the `regions` input.

File: variables.tf (hunk 5 above)
Comment (milesmatthias) [reply variables.tf]: Done

**Review by morgante — CHANGES_REQUESTED** [review]
Sorry, I should have been clearer. I want to remove the `regions` input variable, but we should still return the multi-regional outputs (and multi-regional example)—it can simply return *all* regions instead of having to choose a specific subset.

**Comment by milesmatthias** [comment]
@morgante made that change.

File: README.md (hunk 6 above)
Comment (morgante) [reply README.md]: We need to remove this from the example.

File: README.md [hunk 7]
```diff
@@ -1 +1,80 @@
...
+}
+
+output "region_short_name_map" {
```
Comment (morgante) [inline README.md]: Just outputting the output isn't a very useful example. We should do something like setting a local to `module.utils. region_short_name_map["us-central1"]`.

File: examples/multi_region/outputs.tf [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+output "region_short_name_map" {
+  description = "The 4 or 5 character shortname of any given region."
+  value       = "${module.utils.region_short_name_map}"
+}
+
+output "region_short_name" {
```
Comment (morgante) [inline examples/multi_region/outputs.tf:22]: This output isn't meaningful for this example, so don't include it.

File: outputs.tf [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+
+output "region_short_name_map" {
+  description = "The 4 or 5 character shortname of any given region."
+  value       = "${local.region_short_name_map}"
+}
+
+output "region_short_name" {
+  description = "The 4 or 5 character shortname of a given region."
```
Comment (morgante) [inline outputs.tf]:
```suggestion
  description = "The 4 or 5 character shortname of the region specified in var.region."
```

File: variables.tf [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "region" {
+  description = "A list of GCP region codes for each of the given regions. i.e. \"us-central1\"=\"usc1\""
```
Comment (morgante) [inline variables.tf]: Please give a correct description.

```suggestion
  description = "The GCP region to retrieve a short name for (ex. `us-central1)."
```

**Review by morgante — CHANGES_REQUESTED** [review]

**Comment by milesmatthias** [comment]
@morgante I think this can be merged safely.

**Review by morgante — APPROVED** [review]

**Pull request merged**